[Docs] Describe how action works and depends on each other (#5386)
DevExpress/testcafe: docs/articles/documentation/guides/basic-guides/interact-with-the-page.md

* [Resize Window](#resize-window)
* [Wait](#wait)

> [Test controller](../../reference/test-api/testcontroller/README.md) object includes them as its methods. You can call them in a chained fashion.

Reviewer: `[Test controller](../../reference/test-api/testcontroller/README.md) object includes them as its methods.`
Specify that "them" refers to "these objects"

`You can call them in a chained fashion.`
What do you mean by "chained fashion"?

Author: 1. Is that better:
> [Test controller](../../reference/test-api/testcontroller/README.md) object includes actions as its methods

2. When we call actions in code, we chain them one onto the other with a dot:
Item.**hover**.**click**

Reviewer: Thank you.

1) Test controller object includes actions as its methods
Please add an article to "test controller object". (e.g., "The test controller object")

2) Can we say:
`You can chain these actions when you call them in code.`
OR
`You can chain these actions in code.`

Author: 1. Done

2. Works like a charm, thanks

> Below you can find the list of available actions with reproducible examples and links to their descriptions.

Reviewer: ```suggestion
Below, you can find a list of available actions with reproducible examples and links to their descriptions.
```

* [Populate File Upload Input](../../reference/test-api/testcontroller/setfilestoupload.md)
* [Clear File Upload Input](../../reference/test-api/testcontroller/clearupload.md)

> > The file upload actions allow you to manage the list of files you want to upload. The server receives files after you begin upload, for example, when you [click](../../reference/test-api/testcontroller/click.md) the **Upload** or **Submit** button on a page.

Reviewer: `The server receives files after you begin upload, for example, when you [click](../../reference/test-api/testcontroller/click.md) the **Upload** or **Submit** button on a page.`

Can you clarify what you're trying to say here? For example, do you mean:

`For example, when you [click](../../reference/test-api/testcontroller/click.md) the **Upload** or **Submit** button on a page, the files are uploaded to the server.`

Author: I made an attempt to avoid passive voice here. Apparently, a poor one. There was no intent to change the meaning of this sentence.

Reviewer: Even though "are uploaded to the server" is passive, I think the passive voice is fine in this instance.

});
```

## Work With Iframes

A TestCafe test's [browsing context](https://html.spec.whatwg.org/multipage/browsers.html#windows) is limited to the main window or an `<iframe>`. To use an `<iframe>` in your test,

> switch the context from the main window to this `<iframe>` (and then probably back).

Reviewer: What do you mean by "and then probably back"? For example, do you mean:

`switch the context from the main window to the `<iframe>`. (You may need to switch back to the main window later.)`

Author: Exactly, thanks

Reviewer: Thanks, please apply my suggestion.

Author: done

> If multiple `<iframes>` are present in your test, you should switch between them.

Reviewer: ```suggestion
If your test contains multiple `<iframes>`, you should switch between them.
```

## Handle Native Dialogs

TestCafe allows you to handle native dialogs that the browser may display during the test run.

You can close [alert](https://developer.mozilla.org/en-US/docs/Web/API/Window/alert) and
[beforeunload](https://developer.mozilla.org/en-US/docs/Web/Events/beforeunload) dialogs,
choose an option in [confirm](https://developer.mozilla.org/en-US/docs/Web/API/Window/confirm) dialogs

> or hand over text for [prompt](https://developer.mozilla.org/en-US/docs/Web/API/Window/prompt) dialogs.

Reviewer 2: `or hand over text for...`

I propose using the most simple verbs because most of our customers are not native speakers.
`provide` was a good word.

Reviewer 3: `provide` is banned by correctors currently

Reviewer 3: `provide` is currently banned by correctors

Author: > `provide` is currently banned by correctors

I guess we could use `supply`

Reviewer 2: Ok. Let it be so.

Mouse event | Touch event
----------- | -------------
`mousemove` (when hovering or dragging) | `touchmove` (when dragging)
`mousedown` | `touchstart`
`mouseup`   | `touchend`

## Interaction Requirements

TestCafe actions can interact with elements if they satisfy the following conditions:

> * an element is within the page's [`body`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/body) or in an [`<iframe>`](#work-with-iframes). The element can be outside of a browser window as long as TestCafe can scroll to reach it.

Reviewer 2: * an element is within the [`body`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/body) element (page window or iframe's window). The element can be invisible for the user. TestCafe will try to reach it with a scroll.

Author: edited

> * an element has to be within the page's [`body`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/body) or in an [`<iframe>`](#work-with-iframes). The element may be outside of a browser window- as long as TestCafe can reach it by scrolling.

Reviewer 4: we can change 'has to be' to just 'is' because we have already said that these are conditions that should be satisfied.

* *an element is within...*

And I don't think we need a hyphen in the second sentence. Neither `-ing`.

*The element can be outside of a browser window as long as TestCafe can scroll to reach it.*

Author: I changed all the occurrences

* an element is visible - its properties are as follows:

    Property | Value
    -------- | --------
    `display`  | *not* set to `none`
    `visibility` | set to `visible` (the default value)

>     `width`    | auto calculated as or set to > `0`

Reviewer 2: > auto calculated as or set to > `0`

`>=  1px`.

Author: edited

    `height`   | auto calculated as or set to > `0`

> * a targeted part of the element has to be visible.  

Reviewer 2: > a targeted part of the element has to be visible.

It's not clear. Try to rephrase.

Author: done

> an element isn't obstructed or overlapped.

>     TestCafe actions target a center of an element, or a point specified by the `offsetX` and `offsetY` options of an action. If another element obstructs the target point, the action still executes. The [.click](../../reference/test-api/testcontroller/click.md) action, for example, affects an overlapping element instead of a targeted one.

Reviewer 4: The word `still` is also not quite apt for technical docs. It usually indicates the part with `still` can be shortened or omitted:

*...If another element obstructs the target point, the action is executed with the overlapping element (for instance, the [t.click](../../reference/test-api/testcontroller/click.md) action clicks an overlapping block).*

### Example: Scroll an Element into View

> TestCafe scrolls to reach items that are on the page but not on-screen. Therefore, TestCafe API doesn't have a designated scroll action.

Reviewer 2: I don't like the `designated` word.
Can we change it with `special`, `separate`, etc.?

Author: edited

> dedicated

> TestCafe scrolls to reach items, that are on the page but not on-screen, Therefore, there is no designated scroll action.

Reviewer 4: I guess we don't need a comma before `that` (we never need, in contrast to `which`) and there is a typo in `, Therefore`

Reviewer 4: `there is` and `there are` can also be replaced with more specific wording in tech docs:

*TestCafe scrolls to reach items that are on the page but not on-screen. Therefore, TestCafe API does not have a designated scroll action.*

Dirk also usually changes `has no` to `doesn't have`, but it's British vs American

You can use any action (for example, [hover](#hover)) to scroll towards the desired part of the page.  

> To scroll the page without any action, you can use a [ClientFunction](obtain-client-side-info.md).

Reviewer 2: It's a hack. We should tell about it.
`In rare cases you may need scroll page without using TestCafe's actions. For example <please find out which ones I don't remember>. To scroll the page....`

Author: edited

```js
import { ClientFunction } from 'testcafe';

> const browserscroll = ClientFunction(function() {

Reviewer 2: ```js
const scrollBy = ClientFunction(() => {
    window.scrollBy(0, 1000);
});

test('Test', async t => {
      await scrollBy();
});
```

Author: edited

Mouse event | Touch event
----------- | -------------

> `mousemove` (when hovering or dragging) | `touchmove` (when dragging)

Reviewer: Is it possible to rewrite these without the -ing form?

Author: I've been thinking and can't seem to come up with one
@VasilyStrelyaev any thoughts?

`mousedown` | `touchstart`
`mouseup`   | `touchend`

## Interaction Requirements

TestCafe actions can interact with elements if they satisfy the following conditions:

> * an element is within the page's [`body`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/body) or in an [`<iframe>`](#work-with-iframes). The element can be invisible for the user. If the element is off-screen, TestCafe attempts to reach it with a scroll.

Reviewer: Can we say the following?

`* an element is within the page's [`body`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/body) or in an [`<iframe>`](#work-with-iframes). The element can be invisible to the user. If the element is off-screen, TestCafe uses a scroll action to attempt to locate it.`

Author: Not quite, because this implies that TestCafe has a scroll action, which isn't the case. Full list of actions available in TestCafe is listed [here](https://devexpress.github.io/testcafe/documentation/guides/basic-guides/interact-with-the-page.html)

Does this work:
> an element is within the page's [body](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/body) or in an [<iframe>](#work-with-iframes). The element can be invisible to the user. If the element is off-screen, TestCafe attempts to locate it with a scroll.

Reviewer: It works, thanks. Please apply your suggestion.

Author: done

> * an element is visible - its properties are as follows:

Reviewer: ```suggestion
* an element is visible, with the following properties:
```

    Property | Value
    -------- | --------
    `display`  | *not* set to `none`
    `visibility` | set to `visible` (the default value)

>     `width`    | auto calculated as or set to >= 1px

Reviewer: What do you mean by "as or"?

Author: I meant to say

> **auto-calculated as >= 1px** or **set to >=px**

but wanted to cut the redundancies. I can see the flaw, please advise.

Reviewer: `auto-calculated as >= 1px or set to >=px`

Thank you. Please apply your new suggestion.

Author: done

>     `height`   | auto calculated as or set to >= 1px

Reviewer: See my comment above.

Author: done

> * an element isn't obstructed or overlapped.  

Reviewer: ```suggestion
* an element is not obstructed or overlapped.  
```

>     TestCafe actions target a center of an element, or a point specified by the `offsetX` and `offsetY` options of an action. If another element obstructs the target point, the action is executed with the overlapping element (for instance, the [t.click](../../reference/test-api/testcontroller/click.md) action clicks an overlapping block).

Reviewer: `for instance, the [t.click](../../reference/test-api/testcontroller/click.md) action clicks an overlapping block`

Can you clarify what you mean by "overlapping block"? Is it an overlapped block of text?

Author: It might be a block of text or any element for that matter.

I attempted to avoid repetition. Does this work:

> (for instance, the [t.click](../../reference/test-api/testcontroller/click.md) action clicks an overlapping element).

Reviewer: `(for instance, the t.click action clicks an overlapping element).`

Thanks. Please apply your new suggestion.

### Example: Scroll an Element into View

> TestCafe scrolls to reach items that are on the page but not on-screen. Therefore, TestCafe API doesn't have a dedicated scroll action.

Reviewer: I'm not sure why you used "Therefore". Did you mean to say "However"? For example:

`TestCafe can scroll to reach items that are on the page but not on-screen. However, the TestCafe API does not have a dedicated scroll action.`

Author: This is an attempt to underline causality.

All TestCafe actions incorporate scroll into their routine. **For that reason**, there is no designated scroll action in TestCafe as there is no need for it.

Reviewer: Thanks for the clarification. How about:

`Since TestCafe scrolls to reach items that are on the page but not on-screen, the TestCafe API does not have a dedicated scroll action.`

Author: Good one, edited

`mousemove` (when hovering or dragging) | `touchmove` (dragging only)
`mousedown` | `touchstart`
`mouseup`   | `touchend`

## Interaction Requirements

> To be properly targeted by TestCafe, elements have to satisfy several conditions:

Reviewer 4: We should be very careful when we use adjectives in technical texts.
When writing docs, it's appropriate to ask this boring nerdy question: "Can they be improperly targeted?"
So, we'd better say 

*TestCafe actions can interact with elements if they satisfy the following conditions:*

* an element has to be within the page [`body`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/body) or in an [`<iframe>`](#work-with-iframes).  

* an element has to be visible- neither of its properties should be set to:

    Property | Value
    -------- | --------
    `display`  | `none`
    `visibility` | `hidden`, `collapse`
    `width`    | `0`

>     `height`   | `0`

Reviewer 4: There is a thing with human perception - we are likely to ignore the text and focus on images, code examples, tables. So, many people would skip "neither" and look at the values.

The same problem occurs when we show an example that shouldn't work (to show how not to do). In this instance, we would have to indicate this visually (like with a red background). Otherwise, many will think it's a good example.

There are several ways to address this, including inversion (it brings a little complication because we should take into account the default values):

```
an element has to be visible - its properties should be set as follows:

    Property | Value
     -------- | --------
     `display`  | *not* set to `none`
     `visibility` | set to `visible` (the default value)
     `width`    | auto-calculated as or set to > `0`
     `height`   | auto-calculated as or set to > `0`
```

Author: visibility can be **visible**|hidden|collapse|**initial**|**inherit**; we're looking for the highlighted ones here.
can we write:

```
Property | Value
     -------- | --------
     `visibility` | *not* set to `hidden` or `collapse`
```
